[ADT][NFCI]: Fix iterator category for graph iterators with external …

Maetveis opened the pull request:
…storage

Set the iterator category for graph iterators to input_iterator_tag when the visited set is stored externally. In that case we can't provide multi-pass guarantee, so we should not claim to be a forward iterator.

Fixes: #116400

Maetveis committed:
[ADT][NFCI]: Fix iterator category for graph iterators with external storage

Set the iterator category for graph iterators to input_iterator_tag when
the visited set is stored externally. In that case we can't provide
multi-pass guarantee, so we should not claim to be a forward iterator.

Fixes: #116400

kuhar reviewed (commented):
Can you add a test for this?

Maetveis committed:
[ADT][Test]: multi-Pass safety, iterator_category for graph iterators

Check that using the same iterators for multiple passes produces the
same result, as required by the C++ standard for forward iterators.

Add a static_asserts for the iterator_category of Graph iterators
(BreadthFirstIterator, DepthFirstIterator, PostOrderIterator).

Maetveis commented:
> Can you add a test for this?

Sure, see latest commit.

kazutakahirata reviewed (approved):
LGTM.  Thanks!
